terraform-provider-constellation/internal/provider/cluster_resource.go

```diff
@@ -460,6 +460,19 @@ func (r *ClusterResource) ModifyPlan(ctx context.Context, req resource.ModifyPla
 			"For details, see https://docs.edgeless.systems/constellation/overview/license")
 	}
 
+	// Check microservice version is set to the provider version. // must be done in modify plan to read provider data.
```

Comment:
I don't understand this. Why wouldn't we be able to access `r.providerData` in the creation or config validation?

Reply:
Because TF has its quirks :/ 
Note that Terraform calling the [ValidateResourceConfig](https://developer.hashicorp.com/terraform/plugin/framework/internals/rpcs#validateresourceconfig-rpc) RPC would not call the [ConfigureProvider](https://developer.hashicorp.com/terraform/plugin/framework/internals/rpcs#configureprovider-rpc) RPC first, so implementations need to account for that situation. Configuration validation in Terraform occurs without provider configuration ("offline").

Reply:
We should then remove that validation step to the top of create imo. I think the `ModifyPlan` semantics are quite special, and as a developer, I would not expect validation to happen here. But no strong opinion.